### PR TITLE
Document how to setup a NGINX proxy

### PR DESCRIPTION
**Pull request by toch**
Here the instructions.

It correctly routes and rewrites the requests to the frontend and jupyterhub (including for its websocket). However, I still have an error when running a pipeline in the frontend:

```
[Network error]: TypeError: fetch failed. Backend is unreachable. Is it running?
 ⨯ ApolloError: fetch failed
    at new ApolloError (/code/node_modules/@apollo/client/errors/errors.cjs:33:28)
    at /code/node_modules/@apollo/client/core/core.cjs:2014:78
    at both (/code/node_modules/@apollo/client/utilities/utilities.cjs:1353:31)
    at /code/node_modules/@apollo/client/utilities/utilities.cjs:1344:72
    at new Promise (<anonymous>)
    at Object.then (/code/node_modules/@apollo/client/utilities/utilities.cjs:1344:24)
    at Object.error (/code/node_modules/@apollo/client/utilities/utilities.cjs:1355:49)
    at notifySubscription (/code/node_modules/zen-observable/lib/Observable.js:140:18)
    at onNotify (/code/node_modules/zen-observable/lib/Observable.js:179:3)
    at SubscriptionObserver.error (/code/node_modules/zen-observable/lib/Observable.js:240:7) {
  graphQLErrors: [],
  protocolErrors: [],
  clientErrors: [],
  networkError: TypeError: fetch failed
      at node:internal/deps/undici/undici:13178:13
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
    [cause]: Error: connect ECONNREFUSED 172.18.0.4:8000
        at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1607:16)
        at TCPConnectWrap.callbackTrampoline (node:internal/async_hooks:130:17) {
      errno: -111,
      code: 'ECONNREFUSED',
      syscall: 'connect',
      address: '172.18.0.4',
      port: 8000
    }
  },
  extraInfo: undefined
}
[Network error]: TypeError: fetch failed. Backend is unreachable. Is it running?
 ⨯ ApolloError: fetch failed
    at new ApolloError (/code/node_modules/@apollo/client/errors/errors.cjs:33:28)
    at /code/node_modules/@apollo/client/core/core.cjs:2014:78
    at both (/code/node_modules/@apollo/client/utilities/utilities.cjs:1353:31)
    at /code/node_modules/@apollo/client/utilities/utilities.cjs:1344:72
    at new Promise (<anonymous>)
    at Object.then (/code/node_modules/@apollo/client/utilities/utilities.cjs:1344:24)
    at Object.error (/code/node_modules/@apollo/client/utilities/utilities.cjs:1355:49)
    at notifySubscription (/code/node_modules/zen-observable/lib/Observable.js:140:18)
    at onNotify (/code/node_modules/zen-observable/lib/Observable.js:179:3)
    at SubscriptionObserver.error (/code/node_modules/zen-observable/lib/Observable.js:240:7) {
  graphQLErrors: [],
  protocolErrors: [],
  clientErrors: [],
  networkError: TypeError: fetch failed
      at node:internal/deps/undici/undici:13178:13
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
    [cause]: Error: connect ECONNREFUSED 172.18.0.4:8000
        at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1607:16)
        at TCPConnectWrap.callbackTrampoline (node:internal/async_hooks:130:17) {
      errno: -111,
      code: 'ECONNREFUSED',
      syscall: 'connect',
      address: '172.18.0.4',
      port: 8000
    }
  },
  extraInfo: undefined
}
Error tracking event:  [TypeError: fetch failed] {
  cause:  [Error: connect ECONNREFUSED 172.18.0.4:8000] {
  errno: -111,
  code: 'ECONNREFUSED',
  syscall: 'connect',
  address: '172.18.0.4',
  port: 8000
}
}
```

It seems that it can't reach the pipelines runner, when for the rest it doesn't have any issue.

When checking its container status, it's reported as unhealthy. The Python process is indeed `<defunct>`.

it seems that it crashes after we request to run the pipeline. As I don't have much more logs than that, I don't know where to look at:

```
wait-for-it: waiting 15 seconds for db:5433
wait-for-it: db:5433 is available after 0 seconds
start pipeline runner
Run pipeline: 39b8e3cc-81f0-4fde-bdd8-ea8fae57c8f1
End of run pipeline: 39b8e3cc-81f0-4fde-bdd8-ea8fae57c8f1
```

### CODE REVIEW

**Comment by qgerome** [comment]
Regarding the frontend, does the logs come from the browser console or the container logs ? That's helpful to determine on which end there is a communication problem.

Regarding the pipelines runner, which container is `<defunct>` ? `pipelines_runner` or the container that runs the pipeline ?

**Comment by toch** [comment]
Thanks for asking clarifications:

> Regarding the frontend, does the logs come from the browser console or the container logs ? That's helpful to determine on which end there is a communication problem.

it comes from the container logs. There are also errors in the console logs, here there are

```
13:40:59.233 [GraphQL error]: Message: 'NoneType' object has no attribute 'id', Location: [{"line":2,"column":3}], Path: runPipeline, Extension Code: undefined _app-02ff4da2dabb405c.js:32:2048
13:40:59.235 Uncaught (in promise) ApolloError: 'NoneType' object has no attribute 'id'
    NextJS 51
_app-02ff4da2dabb405c.js:106:6649
```

> Regarding the pipelines runner, which container is <defunct> ? pipelines_runner or the container that runs the pipeline ?

it happens in the `pipelines_runner container`

**Comment by toch** [comment]
As it seems it's a graphql error, I've looked at the app logs, and there is an error at the backend level:

```
'NoneType' object has no attribute 'id'

GraphQL request:2:3
1 | mutation RunWorkspacePipeline($input: RunPipelineInput!) {
2 |   runPipeline(input: $input) {
  |   ^
3 |     success
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/graphql/execution/execute.py", line 521, in execute_field
    result = resolve_fn(source, info, **args)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/code/hexa/user_management/schema.py", line 76, in resolve
    return original_resolver(obj, info, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/code/hexa/pipelines/schema/mutations.py", line 221, in resolve_run_pipeline
    "version_id": str(version.id),
                      ^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'id'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/graphql/execution/execute.py", line 1030, in execute
    result = exe_context.execute_operation(operation, root_value)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/graphql/execution/execute.py", line 353, in execute_operation
    return (
           ^
  File "/usr/local/lib/python3.12/site-packages/graphql/execution/execute.py", line 375, in execute_fields_serially
    result = self.execute_field(
             ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/graphql/execution/execute.py", line 558, in execute_field
    self.handle_field_error(error, return_type)
  File "/usr/local/lib/python3.12/site-packages/graphql/execution/execute.py", line 569, in handle_field_error
    raise error
  File "/usr/local/lib/python3.12/site-packages/graphql/execution/execute.py", line 521, in execute_field
    result = resolve_fn(source, info, **args)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/code/hexa/user_management/schema.py", line 76, in resolve
    return original_resolver(obj, info, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/code/hexa/pipelines/schema/mutations.py", line 221, in resolve_run_pipeline
    "version_id": str(version.id),
                      ^^^^^^^^^^
graphql.error.graphql_error.GraphQLError: 'NoneType' object has no attribute 'id'

GraphQL request:2:3
1 | mutation RunWorkspacePipeline($input: RunPipelineInput!) {
2 |   runPipeline(input: $input) {
  |   ^
3 |     success
"POST /graphql/RunWorkspacePipeline/ HTTP/1.1" 200 1194
"POST /workspaces/credentials/ HTTP/1.1" 200 755
```

**Comment by qgerome** [comment]
**Connection problem in the frontend container**

The frontend container access the openhexa-app API via the browser of the enduser and the nodejs process in the frontend container. The way I understand it, it looks like there is no loopback for this IP (or the container does not have access to it. I'd try to put `http://app:8000` as the `OPENHEXA_BACKEND_URL`.

**Pipelines runner**

So the container `pipelines_runner` is not healthy or does the containers (pipelines) started from the pipelines runner are not healthy ? (Sorry I'm not sure to understand where there is a problem

**Error in the frontend**
I see where it comes from, you have to pull the newest main of openhexa-app. We identified a few bugs with the analytics 🎉  that are fixed on main

**Comment by toch** [comment]
> Connection problem in the frontend container
> 
> The frontend container access the openhexa-app API via the browser of the enduser and the nodejs process in the frontend container. The way I understand it, it looks like there is no loopback for this IP (or the container does not have access to it. I'd try to put http://app:8000 as the OPENHEXA_BACKEND_URL.

This is already what I use, I haven't touched it for the reason you mention. From the `docker inspect` run on the frontend

```
"OPENHEXA_BACKEND_URL=http://app:8000",
```

> Pipelines runner
> 
> So the container pipelines_runner is not healthy or does the containers (pipelines) started from the pipelines runner are not healthy ? (Sorry I'm not sure to understand where there is a problem

it's the runner container itself.

> Error in the frontend
> I see where it comes from, you have to pull the newest main of openhexa-app. We identified a few bugs with the analytics 🎉 that are fixed on main

ok

---

After a live discussion, we found out it was due to
- a not up-to-date version of the app
- a env var that wasn't set correctly

**Comment by toch** [comment]
* [x] rebased on main
* [x] manually tested from scratch locally
* [x] all the last changes integrated

File: script/openhexa.sh [hunk 1]
```diff
@@ -82,6 +83,7 @@ function is_backend_reachable() {
 }
 
 function is_db_accepting_connexion() {
+  # TODO replace with get_config_or_default
```
Comment (qgerome) [inline script/openhexa.sh:86]: Do you want to address this TODO here or in the future ?

**Review by qgerome — APPROVED** [review]
Looks good to me. I just added a comment, feel free to ignore it.

File: script/openhexa.sh (hunk 1 above)
Comment (toch) [reply script/openhexa.sh:86]: later

**Pull request merged**